### PR TITLE
take prewarmed container's memory as used memory

### PR DESCRIPTION
**Pull request by ningyougang**
## Description

Due to this pr (Adjust prewarm container dynamically): https://github.com/apache/openwhisk/pull/4871 is merged
I think it is a good chance to add this feature.

for master branch,  if invoker has `16GB` physical memory, currently, if we configure 12GB user memory and 5GB prewarm pool, invokers will try to create containers up to 17GB memory and it may will end up with OOM.

After apply this feature, user just configure invoker memory to a precise value:  `16GB` , when  reaches `16GB`, it will not create container any more, this can avoid `OOM`

This has another benefit that make user must calculate the prewarmed container's memory carefully, and configure the invoker memory more accurately
```
"CONFIG_whisk_containerPool_userMemory": "{{ hostvars[groups['invokers'][invoker_index | int]].user_memory | default(invoker.userMemory) }}"
```

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by ningyougang** [comment]
Below prs has relation with this pr
* Adjust user memory via api: https://github.com/apache/openwhisk/pull/4917
* Add administrative interface to invoker and controller to reconfigure runtimes https://github.com/apache/openwhisk/pull/4790

`pull/4917` is `adjust userMemory` and `pull/4790` is `adjust prewarmed container`

**Comment by ningyougang** [comment]
upsteam guy: Tyson replied on dev mail list for this pr as below

```
I agree this is a good change, but may require operators to update their configured userMemory and/or their prewarm configs
 
to avoid unexpected problems. It might be good to add some log statements to indicate userMemory available AFTER initial 

prewarm config is applied, to give some hints that userMemory seen at invoker is now lower than before with the same configs?
```
Regarding ` but may require operators to update their configured userMemory and/or their prewarm configs to avoid unexpected problems.`
Currently, below prs has been implemented with `update their configured userMemory and/or their prewarm configs`
* Adjust user memory via api: https://github.com/apache/openwhisk/pull/4917
* Add administrative interface to invoker and controller to reconfigure prewarm config: https://github.com/apache/openwhisk/pull/4790

Regarding `It might be good to add some log statements to indicate userMemory available AFTER initial 
prewarm config is applied, to give some hints that userMemory seen at invoker is now lower than before with the same configs?`
Agree, i think after above two prs(pull/4917 and pull/4790) merged, i can add some logs in this pr after rebased codes.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala [hunk 1]
```diff
@@ -131,9 +131,18 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
           //remove from resent tracking - it may get resent again, or get processed
           resent = None
         }
+        val kind = r.action.exec.kind
+        val memory = r.action.limits.memory.megabytes.MB
+
+        val prewarmedPoolForOtherKind = prewarmedPool.filter { info =>
+          info match {
+            case (_, PreWarmedData(_, `kind`, `memory`, _, _)) => false
```
Comment (style95) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala:139]: Why does it require filtering prewarm pool for other kinds?

Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala]: hm.. i forgot the reason why i want to `get the other kind/memory prewarmpool` ,
but then, i feel use `prewarmedPool` directly is more nature, already updated.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala [hunk 2]
```diff
@@ -131,9 +131,12 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
           //remove from resent tracking - it may get resent again, or get processed
           resent = None
         }
+        val kind = r.action.exec.kind
+        val memory = r.action.limits.memory.megabytes.MB
+
         val createdContainer =
           // Is there enough space on the invoker for this action to be executed.
-          if (hasPoolSpaceFor(busyPool, r.action.limits.memory.megabytes.MB)) {
+          if (hasPoolSpaceFor(busyPool ++ prewarmedPool, memory)) {
```
Comment (style95) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala:139]: I feel this needs proper documentation.

**Review by style95 — APPROVED** [review]
LGTM with a minor nit.

**Comment by style95** [comment]
Merged as it has been around 2 months.

**Pull request merged**